File: src/indexer/languages/rust.ts
import type { SymbolKind } from "../../types.js";
import type { LanguageParser } from "./parser.js";
import {
  buildTreeSitter,
  field,
  named,
  descendants,
  firstDescendant,
  type Node,
  type Emit,
  type Ctx,
} from "./treesitter/base.js";

const isPub = (node: Node): boolean => !!named(node, "visibility_modifier");

const NAMED_ITEM: Record<string, SymbolKind> = {
  function_item: "function",
  function_signature_item: "function",
  struct_item: "class",
  union_item: "class",
  enum_item: "enum",
  trait_item: "interface",
  type_item: "type",
  const_item: "const",
  static_item: "const",
};

const dirOf = (file: string): string => {
  const i = file.lastIndexOf("/");
  return i === -1 ? "" : file.slice(0, i);
};
const baseOf = (file: string): string => {
  const i = file.lastIndexOf("/");
  return i === -1 ? file : file.slice(i + 1);
};

function childModDir(file: string): string {
  const dir = dirOf(file);
  const stem = baseOf(file).replace(/\.rs$/, "");
  if (stem === "mod" || stem === "main" || stem === "lib") return dir;
  return dir ? `${dir}/${stem}` : stem;
}

function ancestorDir(dir: string, ups: number): string {
  let d = dir;
  for (let i = 0; i < ups; i++) d = dirOf(d);
  return d;
}

function crateRootDir(relSet: Set<string>): string {
  for (const r of ["src/lib.rs", "src/main.rs", "lib.rs", "main.rs"]) {
    if (relSet.has(r)) return childModDir(r);
  }
  for (const f of relSet) {
    const b = baseOf(f);
    if (b === "lib.rs" || b === "main.rs") return childModDir(f);
  }
  return "";
}

function modFile(base: string, rest: string[], relSet: Set<string>): string | null {
  if (rest.length === 0) return null;
  const joined = (base ? `${base}/` : "") + rest.join("/");
  for (const cand of [`${joined}.rs`, `${joined}/mod.rs`]) {
    if (relSet.has(cand)) return cand;
  }
  return null;
}

function pathSegs(node: Node): string[] {
  if (!node) return [];
  const t = node.type;
  if (t === "identifier" || t === "type_identifier") return [node.text];
  if (t === "crate" || t === "self" || t === "super") return [t];
  if (t === "scoped_identifier" || t === "scoped_type_identifier") {
    return [...pathSegs(field(node, "path")), ...pathSegs(field(node, "name"))];
  }
  return [];
}

function anchor(
  segs: string[],
  file: string,
  relSet: Set<string>,
): { base: string; rest: string[] } {
  const head = segs[0];
  if (head === "crate") return { base: crateRootDir(relSet), rest: segs.slice(1) };
  if (head === "self") return { base: childModDir(file), rest: segs.slice(1) };
  if (head === "super") {
    let ups = 0;
    while (segs[ups] === "super") ups++;
    return { base: ancestorDir(childModDir(file), ups), rest: segs.slice(ups) };
  }
  return { base: crateRootDir(relSet), rest: segs };
}

function importsForPath(segs: string[], ctx: Ctx, emit: Emit): void {
  if (segs.length === 0) return;
  const { base, rest } = anchor(segs, ctx.file, ctx.relSet);
  const asModule = modFile(base, rest, ctx.relSet);
  if (asModule) emit.import({ from: ctx.file, to: asModule, names: [] });
  const asItem = modFile(base, rest.slice(0, -1), ctx.relSet);
  if (asItem && asItem !== asModule) {
    emit.import({ from: ctx.file, to: asItem, names: [segs[segs.length - 1]] });
  }
}

function modDecl(name: string, ctx: Ctx, emit: Emit): void {
  const to = modFile(childModDir(ctx.file), [name], ctx.relSet);
  if (to) emit.import({ from: ctx.file, to, names: [name] });
}

const isEntryAttr = (attr: string): boolean => /\b(test|bench|no_mangle)\b/.test(attr);
const isCfgTest = (attr: string): boolean => /cfg\s*\(\s*test\s*\)/.test(attr);

function hasExternModifier(node: Node): boolean {
  const mods = named(node, "function_modifiers");
  return !!mods && mods.children.some((c: Node) => c.type === "extern_modifier");
}
function isEntryFn(
  node: Node,
  name: string,
  attrs: string[],
  container: string | undefined,
): boolean {
  if (container === undefined && name === "main") return true;
  if (attrs.some(isEntryAttr)) return true;
  return hasExternModifier(node);
}

function processItem(
  node: Node,
  emit: Emit,
  ctx: Ctx,
  container: string | undefined,
  attrs: string[],
  forceEntry: boolean,
): void {
  const kind = NAMED_ITEM[node.type];
  if (kind) {
    const nameNode = field(node, "name");
    if (!nameNode) return;
    const isFn =
      node.type === "function_item" || node.type === "function_signature_item";
    const isMethod = container !== undefined && isFn;
    const name = isMethod ? `${container}.${nameNode.text}` : nameNode.text;
    const entry = isFn && (forceEntry || isEntryFn(node, nameNode.text, attrs, container));
    emit.symbol({
      name,
      kind: isMethod ? "method" : kind,
      node,
      nameNode,
      exported: isMethod ? false : isPub(node),
      simpleName: isMethod ? nameNode.text : undefined,
      entry: entry || undefined,
    });
    return;
  }
  if (node.type === "impl_item") {
    const typeNode = field(node, "type");
    const typeName =
      typeNode?.type === "type_identifier"
        ? typeNode.text
        : firstDescendant(typeNode ?? node, "type_identifier")?.text;
    const body = field(node, "body");
    processChildren(body ? body.namedChildren : [], emit, ctx, typeName ?? "impl", forceEntry);
    return;
  }
  if (node.type === "mod_item") {
    const body = field(node, "body");
    if (!body) {

      const nn = field(node, "name");
      if (nn) modDecl(nn.text, ctx, emit);
      return;
    }
    const testMod = forceEntry || attrs.some(isCfgTest);
    processChildren(body.namedChildren, emit, ctx, container, testMod);
  }
}

function processChildren(
  children: Node[],
  emit: Emit,
  ctx: Ctx,
  container: string | undefined,
  forceEntry: boolean,
): void {
  let attrs: string[] = [];
  for (const node of children) {
    if (node.type === "attribute_item") {
      attrs.push(node.text);
      continue;
    }
    if (node.type === "line_comment" || node.type === "block_comment") continue;
    processItem(node, emit, ctx, container, attrs, forceEntry);
    attrs = [];
  }
}

function extract(root: Node, emit: Emit, ctx: Ctx): void {

  for (const kind of ["scoped_identifier", "scoped_type_identifier"]) {
    for (const sid of descendants(root, kind)) importsForPath(pathSegs(sid), ctx, emit);
  }
  processChildren(root.namedChildren, emit, ctx, undefined, false);
}

export const rustParser: LanguageParser = {
  name: "rust",
  extensions: ["rs"],

  build: (root, files) => buildTreeSitter(root, files, "rust", extract, { scope: "import" }),
};

File: rust/sens-hook/src/lang/rust.rs
use std::collections::HashSet;

use tree_sitter::Node;

use super::treesitter::{
    declare, descendants, field, first_descendant, named, never_qualified, text, Emitted, Extra,
    Options, Scope,
};

fn item_kind(kind: &str) -> Option<&'static str> {
    match kind {
        "function_item" | "function_signature_item" => Some("function"),
        "struct_item" | "union_item" => Some("class"),
        "enum_item" => Some("enum"),
        "trait_item" => Some("interface"),
        "type_item" => Some("type"),
        "const_item" | "static_item" => Some("const"),
        _ => None,
    }
}

fn named_children_of<'t>(node: &Node<'t>) -> Vec<Node<'t>> {
    let mut cursor = node.walk();
    node.named_children(&mut cursor).collect()
}

fn is_pub(node: &Node) -> bool {
    named(node, "visibility_modifier").is_some()
}

fn dir_of(file: &str) -> &str {
    match file.rfind('/') {
        Some(i) => &file[..i],
        None => "",
    }
}

fn base_of(file: &str) -> &str {
    match file.rfind('/') {
        Some(i) => &file[i + 1..],
        None => file,
    }
}

fn child_mod_dir(file: &str) -> String {
    let dir = dir_of(file);
    let stem = base_of(file).strip_suffix(".rs").unwrap_or(base_of(file));
    if stem == "mod" || stem == "main" || stem == "lib" {
        return dir.to_string();
    }
    if dir.is_empty() {
        stem.to_string()
    } else {
        format!("{dir}/{stem}")
    }
}

fn ancestor_dir(dir: &str, ups: usize) -> String {
    let mut d = dir.to_string();
    for _ in 0..ups {
        d = dir_of(&d).to_string();
    }
    d
}

fn crate_root_dir(rel_set: &HashSet<String>) -> String {
    for candidate in ["src/lib.rs", "src/main.rs", "lib.rs", "main.rs"] {
        if rel_set.contains(candidate) {
            return child_mod_dir(candidate);
        }
    }
    for f in rel_set {
        let b = base_of(f);
        if b == "lib.rs" || b == "main.rs" {
            return child_mod_dir(f);
        }
    }
    String::new()
}

fn mod_file(base: &str, rest: &[String], rel_set: &HashSet<String>) -> Option<String> {
    if rest.is_empty() {
        return None;
    }
    let prefix = if base.is_empty() { String::new() } else { format!("{base}/") };
    let joined = format!("{prefix}{}", rest.join("/"));
    for cand in [format!("{joined}.rs"), format!("{joined}/mod.rs")] {
        if rel_set.contains(&cand) {
            return Some(cand);
        }
    }
    None
}

fn path_segs(node: &Node, source: &str) -> Vec<String> {
    match node.kind() {
        "identifier" | "type_identifier" => vec![text(node, source).to_string()],
        "crate" | "self" | "super" => vec![node.kind().to_string()],
        "scoped_identifier" | "scoped_type_identifier" => {
            let mut out = Vec::new();
            if let Some(path) = field(node, "path") {
                out.extend(path_segs(&path, source));
            }
            if let Some(name) = field(node, "name") {
                out.extend(path_segs(&name, source));
            }
            out
        }
        _ => Vec::new(),
    }
}

fn anchor(segs: &[String], file: &str, rel_set: &HashSet<String>) -> (String, Vec<String>) {
    match segs.first().map(String::as_str) {
        Some("crate") => (crate_root_dir(rel_set), segs[1..].to_vec()),
        Some("self") => (child_mod_dir(file), segs[1..].to_vec()),
        Some("super") => {
            let ups = segs.iter().take_while(|s| *s == "super").count();
            (ancestor_dir(&child_mod_dir(file), ups), segs[ups..].to_vec())
        }
        _ => (crate_root_dir(rel_set), segs.to_vec()),
    }
}

fn imports_for_path(segs: &[String], file: &str, rel_set: &HashSet<String>, out: &mut Emitted) {
    if segs.is_empty() {
        return;
    }
    let (base, rest) = anchor(segs, file, rel_set);
    let as_module = mod_file(&base, &rest, rel_set);
    if let Some(m) = &as_module {
        out.import(file, m.clone(), Vec::new());
    }
    let parent = if rest.is_empty() { Vec::new() } else { rest[..rest.len() - 1].to_vec() };
    if let Some(as_item) = mod_file(&base, &parent, rel_set) {
        if Some(&as_item) != as_module.as_ref() {
            let last = segs.last().cloned().unwrap_or_default();
            out.import(file, as_item, vec![last]);
        }
    }
}

fn mod_decl(name: &str, file: &str, rel_set: &HashSet<String>, out: &mut Emitted) {
    if let Some(to) = mod_file(&child_mod_dir(file), &[name.to_string()], rel_set) {
        out.import(file, to, vec![name.to_string()]);
    }
}

fn word_in(haystack: &str, word: &str) -> bool {
    haystack
        .split(|c: char| !(c.is_alphanumeric() || c == '_'))
        .any(|w| w == word)
}

fn is_entry_attr(attr: &str) -> bool {
    word_in(attr, "test") || word_in(attr, "bench") || word_in(attr, "no_mangle")
}

fn is_cfg_test(attr: &str) -> bool {
    let compact: String = attr.chars().filter(|c| !c.is_whitespace()).collect();
    compact.contains("cfg(test)")
}

fn has_extern_modifier(node: &Node) -> bool {
    named(node, "function_modifiers").is_some_and(|mods| {
        let mut cursor = mods.walk();
        mods.children(&mut cursor).any(|c| c.kind() == "extern_modifier")
    })
}

fn is_entry_fn(node: &Node, name: &str, attrs: &[String], container: Option<&str>) -> bool {
    if container.is_none() && name == "main" {
        return true;
    }
    if attrs.iter().any(|a| is_entry_attr(a)) {
        return true;
    }
    has_extern_modifier(node)
}

fn process_item(
    node: &Node,
    source: &str,
    file: &str,
    rel_set: &HashSet<String>,
    container: Option<&str>,
    attrs: &[String],
    force_entry: bool,
    out: &mut Emitted,
) {
    if let Some(kind) = item_kind(node.kind()) {
        let Some(name_node) = field(node, "name") else { return };
        let is_fn = node.kind() == "function_item" || node.kind() == "function_signature_item";
        let is_method = container.is_some() && is_fn;
        let simple = text(&name_node, source).to_string();
        let name = match container {
            Some(c) if is_method => format!("{c}.{simple}"),
            _ => simple.clone(),
        };
        let entry = is_fn && (force_entry || is_entry_fn(node, &simple, attrs, container));
        declare(
            out,
            source,
            name,
            if is_method { "method" } else { kind },
            node,
            &name_node,
            if is_method { false } else { is_pub(node) },
            Extra {
                simple_name: if is_method { Some(simple) } else { None },
                entry,
                ..Extra::default()
            },
        );
        return;
    }

    if node.kind() == "impl_item" {
        let type_node = field(node, "type");
        let type_name = match type_node {
            Some(t) if t.kind() == "type_identifier" => Some(text(&t, source).to_string()),
            Some(t) => first_descendant(&t, "type_identifier").map(|n| text(&n, source).to_string()),
            None => first_descendant(node, "type_identifier").map(|n| text(&n, source).to_string()),
        };
        let body = field(node, "body");
        let children = body.map(|b| named_children_of(&b)).unwrap_or_default();
        let container = type_name.unwrap_or_else(|| "impl".to_string());
        process_children(&children, source, file, rel_set, Some(&container), force_entry, out);
        return;
    }

    if node.kind() == "mod_item" {
        let Some(body) = field(node, "body") else {
            if let Some(nn) = field(node, "name") {
                mod_decl(text(&nn, source), file, rel_set, out);
            }
            return;
        };
        let test_mod = force_entry || attrs.iter().any(|a| is_cfg_test(a));
        let children = named_children_of(&body);
        process_children(&children, source, file, rel_set, container, test_mod, out);
    }
}

fn process_children(
    children: &[Node],
    source: &str,
    file: &str,
    rel_set: &HashSet<String>,
    container: Option<&str>,
    force_entry: bool,
    out: &mut Emitted,
) {
    let mut attrs: Vec<String> = Vec::new();
    for node in children {
        if node.kind() == "attribute_item" {
            attrs.push(text(node, source).to_string());
            continue;
        }
        if node.kind() == "line_comment" || node.kind() == "block_comment" {
            continue;
        }
        process_item(node, source, file, rel_set, container, &attrs, force_entry, out);
        attrs.clear();
    }
}

pub fn extract(root: &Node, source: &str, file: &str, rel_set: &HashSet<String>, out: &mut Emitted) {
    for kind in ["scoped_identifier", "scoped_type_identifier"] {
        for sid in descendants(root, kind) {
            imports_for_path(&path_segs(&sid, source), file, rel_set, out);
        }
    }
    let children = named_children_of(root);
    process_children(&children, source, file, rel_set, None, false, out);
}

pub fn options() -> Options {
    Options { scope: Scope::Import, qualified_use: never_qualified }
}
